### PR TITLE
Feature/revamp/email based customers

### PR DESCRIPTION
**Pull request by fivetran-jamie**
**Are you a current Fivetran customer?** 

fivetran made PR

**What change(s) does this PR introduce?** 

adds new end models for customers that are rolled up to the EMAIL instead of customer_id

it does NOT include full documentation. this will be completed in a later PR

**Did you update the CHANGELOG?** 

- [x] Yes

**Does this PR introduce a breaking change?**

- [ ] Yes (please provide breaking change details below.)
- [x] No  (please provide an explanation as to how the change is non-breaking below.) additional models

**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 

- [ ] Yes

**Is this PR in response to a previously created Bug or Feature Request**

- [ ] Yes, Issue/Feature https://github.com/fivetran/dbt_shopify/issues/37
- [ ] No 

**How did you test the PR changes?** 

 

- [ ] Buildkite  
- [x] Local (please provide additional testing details below)

**Select which warehouse(s) were used to test the PR**
 

- [ ] BigQuery
- [ ] Redshift
- [ ] Snowflake
- [ ] Postgres
- [ ] Databricks
- [ ] Other (provide details below)

**Provide an emoji that best describes your current mood**
 
🌧️ 

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

### CODE REVIEW

File: packages.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ packages:
 #   version: [">=0.7.0", "<0.8.0"]
 
   - git: https://github.com/fivetran/dbt_shopify_source.git
-    revision: feature/package-revamp
+    revision: feature/revamp/revert-timestamp-name-change
```
Comment (fivetran-jamie) [inline packages.yml:6]: https://github.com/fivetran/dbt_shopify_source/pull/42

**Comment by fivetran-jamie** [comment]
buildkite may not pass on other models because i have not yet propagated source-package-renames and -deprecations to the transform models

File: packages.yml (hunk 1 above)
Comment (fivetran-avinash) [reply packages.yml:6]: The usual "make sure to change the revision" when merging =)

Comment (fivetran-avinash) [reply packages.yml:6]: The usual "make sure to change the revision" when merging =)

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 ## 🎉 Documentation and Feature Updates
 - Updated README documentation updates for easier navigation and setup of the dbt package
 - Included `shopify_[source_table_name]_identifier` variable within the Shopify source package for additional flexibility within the package when source tables are named differently.
+- **New model alert**: The package now includes customer models that are based on _email_ rather than _customer_id_ ([PR #45](https://github.com/fivetran/dbt_shopify/pull/45)):
+  - `shopify__customer_emails`
+  - `shopify__customer_email_cohorts`
+  - Intermediate models that roll customer_ids up to emails
```
Comment (fivetran-avinash) [inline CHANGELOG.md:8]: It's probably worth naming the intermediate models!

Also it's unclear as to why the seed files are added--all the data looks fine but might want to mention why these seed files are now being added?

File: models/intermediate/shopify__emails__order_aggregates.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    select *
+    from {{ var('shopify_order') }}
+    where email is not null
+
+), transactions as (
+
+    select *
+    from {{ ref('shopify__transactions' )}}
+    where lower(status) = 'success'
```
Comment (fivetran-avinash) [inline models/intermediate/shopify__emails__order_aggregates.sql:11]: I'm wondering if this where condition should be configurable? It makes sense in most cases for this to hold true but maybe customers will sometimes want to have transactions in other states too count toward their order definitions. Probably not needed now but could be a good feature request to file for later.

File: models/intermediate/intermediate.yml [hunk 4]
```diff
@@ -13,4 +13,16 @@ models:
           combination_of_columns:
             - order_id
             - source_relation
-  - name: shopify__orders__order_refunds
\ No newline at end of file
+  - name: shopify__orders__order_refunds
```
Comment (fivetran-avinash) [inline models/intermediate/intermediate.yml:16]: doesn't look related to this task, but wonder why doesn't `shopify__orders__order_refunds` have its own `dbt_utils.unique_combination_of_columns` test.

File: models/intermediate/shopify__customer_email_rollup.sql [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                                'is_tax_exempt', 'is_verified_email'] %}
+        {% for col in cols %}
+            {% if col.column|lower not in except_cols %}
+            , max(case when customer_index = 1 then {{ col.column }} else null end) as {{ col.column }}
+            {% endif %}
+        {% endfor %}
+
+    from customers 
+
+    group by 1,2
```
Comment (fivetran-avinash) [inline models/intermediate/shopify__customer_email_rollup.sql:48]: Aren't we supposed to use the `dbt_utils` group by function? Or did we go with the standard formatting because the migration isn't complete yet?

File: models/intermediate/shopify__customer_email_rollup.sql [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        {{ fivetran_utils.max_bool("is_verified_email") }} as is_verified_email
+
+        -- for all other fields, just take the latest value
+        {% set cols = adapter.get_columns_in_relation(ref('stg_shopify__customer')) %}
+        {% set except_cols = ['_fivetran_synced', 'email', 'source_relation', 'customer_id', 'phone', 'created_at', 
+                                'updated_at', 'has_accepted_marketing', 'accepts_marketing_updated_at', 'orders_count', 'total_spent',
+                                'is_tax_exempt', 'is_verified_email'] %}
+        {% for col in cols %}
+            {% if col.column|lower not in except_cols %}
+            , max(case when customer_index = 1 then {{ col.column }} else null end) as {{ col.column }}
```
Comment (fivetran-avinash) [inline models/intermediate/shopify__customer_email_rollup.sql:42]: Commas is supposed to be trailing I think?

File: models/shopify.yml [hunk 7]
```diff
@@ -251,6 +282,61 @@ models:
...
+        description: The total amount of money that the customer has been refunded on orders across their order history.
+      - name: lifetime_total_amount
+        description: The total amount of money (minus refunds) that the customer has spent across their order history.
+      - name: first_account_created_at
+        description: Timestamp of when the first account associated with this email was created.
+      - name: last_account_created_at
+        description: Timestamp of when the last account associated with this email was created.
+      - name: accepts_marketing_last_updated_at
+        description: When the customer consented or objected to receiving marketing material by email. 
+
```
Comment (fivetran-avinash) [inline models/shopify.yml:339]: The `shopify__order_lines` fields need to be added it looks like (looks like they were never present there).

File: models/shopify__order_lines.sql [hunk 8]
```diff
@@ -58,8 +58,10 @@ with order_lines as (
         product_variants.option_1 as variant_option_1,
         product_variants.option_2 as variant_option_2,
         product_variants.option_3 as variant_option_3,
-        product_variants.tax_code as variant_tax_code,
-        product_variants.is_requiring_shipping as variant_is_requiring_shipping
+        product_variants.tax_code as variant_tax_code
+        {# ,
+        use inventoryitem or order line itself
```
Comment (fivetran-avinash) [inline models/shopify__order_lines.sql:63]: what is this line doing?

File: models/shopify__customer_email_cohorts.sql [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+), customer_calendar as (
+
+    select
+        calendar.date_day as date_month,
+        customers.email,
+        customers.first_order_timestamp,
+        customers.source_relation,
+        {{ dbt.date_trunc('month', 'first_order_timestamp') }} as cohort_month
+    from calendar
+    inner join customers
```
Comment (fivetran-avinash) [inline models/shopify__customer_email_cohorts.sql:26]: Couldn't the `inner join` cause gaps in reporting if there are months with no customer email data? Unlikely but possible.

File: models/shopify.yml [hunk 10]
```diff
@@ -32,6 +32,37 @@ models:
...
 
+  - name: shopify__customer_email_cohorts
+    description: Each record represents a customer's (on the EMAIL grain) performance in a calendar month.
+    columns:
+      - name: cohort_month
+        description: The month the cohort belongs to, i.e the first month the customer had an order.
+      - name: cohort_month_number
+        description: The 'number' of the `date_month` of the record, i.e. how many months from their start month this cohort occurred
+      - name: customer_cohort_id
+        description: Unique key representing a customer in a given month.
```
Comment (fivetran-avinash) [inline models/shopify.yml:43]: Would probably mention the fields used to create the `cohort_id` (also it doesn't look like `source_relation` is mentioned here even though it's brought into the final model.

Comment (fivetran-avinash) [inline models/shopify.yml:43]: Would probably mention the fields used to create the `cohort_id` (also it doesn't look like `source_relation` is mentioned here even though it's brought into the final model.

File: integration_tests/dbt_project.yml [hunk 11]
```diff
@@ -83,4 +107,55 @@ seeds:
       +column_types:
         id: "{%- if target.type == 'bigquery' -%} INT64 {%- else -%} bigint {%- endif -%}"
         order_id: "{%- if target.type == 'bigquery' -%} INT64 {%- else -%} bigint {%- endif -%}"
-        user_id: "{%- if target.type == 'bigquery' -%} INT64 {%- else -%} bigint {%- endif -%}"
\ No newline at end of file
+        user_id: "{%- if target.type == 'bigquery' -%} INT64 {%- else -%} bigint {%- endif -%}"
+    shopify_abandoned_checkout_discount_code_data:
```
Comment (fivetran-avinash) [inline integration_tests/dbt_project.yml:111]: what were the seed files needed for in this PR? is this data being brought up into the `orders` and `customers` models and needed for testing across DBs?

File: models/intermediate/shopify__customer_email_rollup.sql [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        sum(total_spent) as total_spent,
+
+        -- take true if ever given for boolean fields
+        {{ fivetran_utils.max_bool("has_accepted_marketing") }} as has_accepted_marketing,
+        {{ fivetran_utils.max_bool("case when customer_index = 1 then is_tax_exempt else null end") }} as is_tax_exempt, -- since this changes every year
+        {{ fivetran_utils.max_bool("is_verified_email") }} as is_verified_email
+
+        -- for all other fields, just take the latest value
+        {% set cols = adapter.get_columns_in_relation(ref('stg_shopify__customer')) %}
+        {% set except_cols = ['_fivetran_synced', 'email', 'source_relation', 'customer_id', 'phone', 'created_at',
```
Comment (fivetran-avinash) [inline models/intermediate/shopify__customer_email_rollup.sql:37]: the below for loop is grabbing the latest value of all columns NOT in this array?

File: models/shopify__customer_email_cohorts.sql [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        customer_calendar.source_relation,
+        coalesce(count(distinct orders.order_id), 0) as order_count_in_month,
+        coalesce(sum(orders.order_adjusted_total), 0) as total_price_in_month,
+        coalesce(sum(orders.line_item_count), 0) as line_item_count_in_month
+    from customer_calendar
+    left join orders
+        on customer_calendar.email = orders.email
+        and customer_calendar.source_relation = orders.source_relation
+        and customer_calendar.date_month = cast({{ dbt.date_trunc('month', 'created_timestamp') }} as date)
+    group by 1,2,3,4,5
```
Comment (fivetran-avinash) [inline models/shopify__customer_email_cohorts.sql:45]: similar question on group bys and dbt_utils

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-jamie Hey Jamie! Sorry for the delay, trying to wrap my head around some of the intricacies of the new changes. I think it mostly looks good with some clarifying questions and additional documentation needed. Let me know your thoughts and it should be good for approval by tomorrow.

File: CHANGELOG.md (hunk 2 above)
Comment (fivetran-jamie) [reply CHANGELOG.md:8]: seed data is added because of the new models i added in the [source package](https://github.com/fivetran/dbt_shopify_source/pull/39) -- in the later documentation PR, i'll delineate all the new source models that were added and note that they have seed data

File: models/intermediate/shopify__emails__order_aggregates.sql (hunk 3 above)
Comment (fivetran-jamie) [reply models/intermediate/shopify__emails__order_aggregates.sql:11]: yeah maybe for later, as i totally copied this logic from the pre-existing `shopify__customers__order_aggregates` model

Comment (fivetran-jamie) [reply models/intermediate/shopify__emails__order_aggregates.sql:11]: the other values for status are pending, failure, and error, so i think it still makes sense to limit to successful transactions for this particular metric. i could however see a separate metric around "failed transactions" being valuable

File: models/intermediate/intermediate.yml (hunk 4 above)
Comment (fivetran-jamie) [reply models/intermediate/intermediate.yml:16]: not sure, i'd consider this out of scope for this task tho

File: models/intermediate/shopify__customer_email_rollup.sql (hunk 5 above)
Comment (fivetran-jamie) [reply models/intermediate/shopify__customer_email_rollup.sql:48]: if i remember correctly, we agreed to use the macro when grouping by 3 or more fields?

File: models/intermediate/shopify__customer_email_rollup.sql (hunk 6 above)
Comment (fivetran-jamie) [reply models/intermediate/shopify__customer_email_rollup.sql:42]: because it's at the end of the select statement it's a lot cleaner to use leading IMO (otherwise we'd need to conditionally add a comma to the end of the previous field), open to opinions however

File: CHANGELOG.md (hunk 2 above)
Comment (fivetran-jamie) [reply CHANGELOG.md:8]: and yah i agree on the intermediate models -- maybe an expandable/collapsable dropdown for those ones would be good 🤔

File: integration_tests/dbt_project.yml (hunk 11 above)
Comment (fivetran-jamie) [reply integration_tests/dbt_project.yml:111]: the seed files were just for the new tables. because this PR is pointing to my working branch of the source package, Buildkite expected those new tables when running the upstream source package stuff

Comment (fivetran-jamie) [reply integration_tests/dbt_project.yml:111]: so the seeds aren't super relevant, this just came up because this is my first PR on the transform package for this project

File: models/intermediate/shopify__customer_email_rollup.sql (hunk 12 above)
Comment (fivetran-jamie) [reply models/intermediate/shopify__customer_email_rollup.sql:37]: yep since i do this outside of the for-loop, don't want duplicates

File: models/shopify.yml (hunk 10 above)
Comment (fivetran-jamie) [reply models/shopify.yml:43]: agree and will add source_relation

File: models/shopify.yml (hunk 7 above)
Comment (fivetran-jamie) [reply models/shopify.yml:339]: i believe they're on line [383](https://github.com/fivetran/dbt_shopify/pull/45/files#diff-ca3b92d02c70e6d6d0d754cc475c032da6910c18ffeafaecda37eb7251ba2aedR383), unless you meant certain fields in particular?

File: models/shopify__customer_email_cohorts.sql (hunk 9 above)
Comment (fivetran-jamie) [reply models/shopify__customer_email_cohorts.sql:26]: yeah that'd be an issue if we were joining using a `=`, but since we are using `<=` this inner join should eliminate the appropriate chunks of data, i think

File: models/intermediate/shopify__customer_email_rollup.sql (hunk 5 above)
Comment (fivetran-avinash) [reply models/intermediate/shopify__customer_email_rollup.sql:48]: Ah, wasn't really sure, it wasn't documented in the style guide.

File: models/shopify__customer_email_cohorts.sql (hunk 13 above)
Comment (fivetran-jamie) [reply models/shopify__customer_email_cohorts.sql:45]: this one i should use indeed group_by for

File: models/shopify__order_lines.sql (hunk 8 above)
Comment (fivetran-jamie) [reply models/shopify__order_lines.sql:63]: just commenting out a deprecated field and making a note to bring it in from `inventoryitem` instead (for a later PR)

File: models/shopify.yml (hunk 7 above)
Comment (fivetran-avinash) [reply models/shopify.yml:339]: @fivetran-jamie The `variant_tax_code` and `variant_is_requiring_shipping` fields aren't documented in that model.

Comment (fivetran-avinash) [reply models/shopify.yml:339]: Oh `variant_is_requiring_shipping` is a dead field per your comment, guess you don't have to bring that one in!

File: models/intermediate/shopify__emails__order_aggregates.sql (hunk 3 above)
Comment (fivetran-jamie) [reply models/intermediate/shopify__emails__order_aggregates.sql:11]: i'll note for my "enhance existing models" task to perhaps add such metrics

File: models/shopify__customer_email_cohorts.sql (hunk 9 above)
Comment (fivetran-avinash) [reply models/shopify__customer_email_cohorts.sql:26]: I think if `customers` don't have any date data for a month (like say customers had data from October 2022 and December 2022 but not November 2022), then there would be no email cohort data from November because of the inner join rather than the left join since the inner join wouldn't recognize null dates.  

(Unless the `customers` table also has all the dates by default, in which case disregard). 

So in the cohort data you'd skip over a month because it'd be all null. When you'd probably just want zeroes to show up. Just want to make sure!

File: integration_tests/dbt_project.yml (hunk 11 above)
Comment (fivetran-avinash) [reply integration_tests/dbt_project.yml:111]: ah got it, thanks for clarifying!

File: models/intermediate/shopify__customer_email_rollup.sql (hunk 6 above)
Comment (fivetran-avinash) [reply models/intermediate/shopify__customer_email_rollup.sql:42]: sounds good to me!

File: models/shopify__order_lines.sql (hunk 8 above)
Comment (fivetran-avinash) [reply models/shopify__order_lines.sql:63]: Got it, since it's a note should we comment it out?

File: models/shopify.yml (hunk 10 above)
Comment (fivetran-jamie) [reply models/shopify.yml:43]: hm source relation is not included in the ymls for any model (source or transform). adding this everywhere will fall into my larger documentation task/PR later (noted this in the height task)

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-jamie Thanks for the prompt re-review! 

For some reason it's telling me to submit another review rather than just submitting comment responses, but here are the follow-ups.

File: models/shopify.yml (hunk 7 above)
Comment (fivetran-jamie) [reply models/shopify.yml:339]: ah gotcha i will handle this in the documentation PR

File: models/shopify__customer_email_cohorts.sql (hunk 9 above)
Comment (fivetran-jamie) [reply models/shopify__customer_email_cohorts.sql:26]: ah so this join looks at the customers _first_ order timestamp. so if the customer made an order in october, but didn't make another order until december, november would still show up, as it is later than the customer's first order-timestamp

File: models/shopify__order_lines.sql (hunk 8 above)
Comment (fivetran-jamie) [reply models/shopify__order_lines.sql:63]: ah yeah `{# ... #}` is the commenting syntax for sql-jinja! same as `/* ... */`, except this works for jinja as well

File: models/shopify__customer_email_cohorts.sql (hunk 9 above)
Comment (fivetran-avinash) [reply models/shopify__customer_email_cohorts.sql:26]: Ahhhh nice! Yeah then I think the inner join should be good then.

File: models/shopify__order_lines.sql (hunk 8 above)
Comment (fivetran-avinash) [reply models/shopify__order_lines.sql:63]: got it! weird that the syntax in github makes it look like code though

**Review by fivetran-avinash — APPROVED** [review]
lgtm!
